src/database.py

```diff
@@ -0,0 +1,144 @@
+from ops.framework import Object
+from psycopg2 import sql
+from psycopg2.extensions import connection
+
+logger = logging.getLogger(__name__)
+
+
+class DatabaseObserver(Object):
+    """The Database relation observer."""
+
+    _RELATION_NAME = "database"
+
+    def __init__(self, charm: CharmBase):
+        """Initialize the observer and register event handlers.
+
+        Args:
+            charm: The parent charm to attach the observer to.
+        """
+        super().__init__(charm, "database-observer")
+        self._charm = charm
+        # SUPERUSER is required to update pg_database
+        self.database = DatabaseRequires(
+            self._charm,
+            relation_name=self._RELATION_NAME,
+            database_name=self._charm.app.name,
+            extra_user_roles="SUPERUSER",
+        )
+
+    def get_relation_data(self) -> typing.Optional[typing.Dict]:
+        """Get database data from relation.
+
+        Returns:
+            Dict: Information needed for setting environment variables.
+        """
+        if self.model.get_relation(self._RELATION_NAME) is None:
+            return None
+
+        relation_id = self.database.relations[0].id
+        relation_data = self.database.fetch_relation_data()[relation_id]
+
+        endpoint = relation_data.get("endpoints", ":")
+
+        return {
+            "POSTGRES_USER": relation_data.get("username", ""),
+            "POSTGRES_PASSWORD": relation_data.get("password", ""),
+            "POSTGRES_HOST": endpoint.split(":")[0],
+            "POSTGRES_PORT": endpoint.split(":")[1],
+            "POSTGRES_DB": self._charm.app.name,
+        }
+
+    def prepare_database(self) -> None:
+        """Change database collate and ctype as required by Synapse.
+
+        Raises:
+           Error: something went wrong while preparing the database.
+        """
+        conn = self.get_conn()
+        database_name = self.get_database_name()
+        if conn is not None:
+            try:
+                with conn.cursor() as curs:
+                    curs.execute(
+                        "UPDATE pg_database "
+                        "SET datcollate='C', datctype='C' "
+                        "WHERE datname=%s",
+                        (database_name,),
+                    )
+            except psycopg2.Error as exc:
+                logger.error("Failed to prepare database: %s", str(exc))
+                raise
+
+    def erase_database(self) -> None:
+        """Erase database.
+
+        Raises:
+           Error: something went wrong while erasing the database.
+        """
+        conn = self.get_conn("template1")
+        database_name = self.get_database_name()
+        if conn is not None:
+            try:
+                with conn.cursor() as curs:
+                    curs.execute(sql.SQL("DROP DATABASE {}").format(sql.Identifier(database_name)))
+                    curs.execute(
+                        sql.SQL(
+                            "CREATE DATABASE {} "
```

Comment:
Similar to above, is this vulnerable to SQL injection?

Reply:
This is another way of doing the same thing as the UPDATE query, not vulnerable to SQL injection. I'll change the previous one to follow the same pattern.

Reply:
Now all queries are being constructed via sql.SQL